llvm/lib/Object/ELF.cpp

```diff
@@ -719,60 +730,83 @@ decodeBBAddrMapImpl(const ELFFile<ELFT> &EF,
-        return createError("failed to get relocation data for offset: " +
-                           Twine::utohexstr(SectionOffset) + " in section " +
-                           describe(EF, Sec));
-      }
-      Address = FOTIterator->second;
-    }
-    uint32_t NumBlocks = readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr);
-
+    uint32_t NumBlocksInBBRange = 0;
+    uint32_t NumBBRanges = 1;
+    typename ELFFile<ELFT>::uintX_t RangeBaseAddress = 0;
     std::vector<BBAddrMap::BBEntry> BBEntries;
-    uint32_t PrevBBEndOffset = 0;
-    for (uint32_t BlockIndex = 0;
-         !MetadataDecodeErr && !ULEBSizeErr && Cur && (BlockIndex < NumBlocks);
-         ++BlockIndex) {
-      uint32_t ID = Version >= 2
-                        ? readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr)
-                        : BlockIndex;
-      uint32_t Offset = readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr);
-      uint32_t Size = readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr);
-      uint32_t MD = readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr);
-      if (Version >= 1) {
-        // Offset is calculated relative to the end of the previous BB.
-        Offset += PrevBBEndOffset;
-        PrevBBEndOffset = Offset + Size;
+    if (FeatEnable.MultiBBRange) {
+      NumBBRanges = readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr);
+    } else {
+      uint64_t RelocationOffsetInSection = Cur.tell();
+      RangeBaseAddress =
+          static_cast<typename ELFFile<ELFT>::uintX_t>(Data.getAddress(Cur));
+      if (!Cur)
+        return Cur.takeError();
+      if (IsRelocatable) {
+        Expected<unsigned> AddressOrErr =
+            GetAddressForRelocation(RelocationOffsetInSection);
+        if (!AddressOrErr)
+          return AddressOrErr.takeError();
+        RangeBaseAddress = *AddressOrErr;
       }
-      Expected<BBAddrMap::BBEntry::Metadata> MetadataOrErr =
-          BBAddrMap::BBEntry::Metadata::decode(MD);
-      if (!MetadataOrErr) {
-        MetadataDecodeErr = MetadataOrErr.takeError();
-        break;
+      NumBlocksInBBRange = readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr);
+    }
+    std::vector<BBAddrMap::BBRangeEntry> BBRangeEntries;
+    uint32_t TotalNumBlocks = 0;
+    for (uint32_t BBRangeIndex = 0; BBRangeIndex < NumBBRanges;
+         ++BBRangeIndex) {
+      uint32_t PrevBBEndOffset = 0;
+      if (FeatEnable.MultiBBRange) {
+        uint64_t RelocationOffsetInSection = Cur.tell();
+        RangeBaseAddress =
+            static_cast<typename ELFFile<ELFT>::uintX_t>(Data.getAddress(Cur));
+        if (IsRelocatable) {
+          assert(RangeBaseAddress == 0);
+          Expected<unsigned> AddressOrErr =
+              GetAddressForRelocation(RelocationOffsetInSection);
+          if (!AddressOrErr)
+            return AddressOrErr.takeError();
+          RangeBaseAddress = *AddressOrErr;
+        }
+        NumBlocksInBBRange = readULEB128As<uint32_t>(Data, Cur, ULEBSizeErr);
```

Comment:
This seems nearly the same as at line 751. The differences seem to be this has an assert on the RangeBaseAddress and 751 short circuits on Cur. If it is possible, could it be moved into a lambda so that this and 751 can just call the lambda?